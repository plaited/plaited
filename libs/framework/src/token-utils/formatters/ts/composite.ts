import { camelCase } from '@plaited/utils'
import { Formatter, CompositeToken } from '../../types.js'
import { hasAlias } from '../has-alias.js'
import { resolveTSVar, isValidAlias } from '../ts-utils.js'

export const composite: Formatter<CompositeToken> = (token, { tokenPath, allTokens }) => {
  const { $value } = token
  if (hasAlias($value) && !isValidAlias($value, allTokens)) {
    return ''
  }
  if (hasAlias($value)) {
    return `export const ${camelCase(tokenPath.join(' '))} = ${resolveTSVar($value, allTokens)}`
  }
  const toRet: string[] = []
  for (const key in $value) {
    const val = $value[key]
    if (isValidAlias(val, allTokens)) {
      toRet.push(`  ${key}: ${resolveTSVar(val, allTokens)},`)
    }
  }
  return [`export const ${camelCase(tokenPath.join(' '))} = {`, ...toRet, '}'].join('\n')
}
