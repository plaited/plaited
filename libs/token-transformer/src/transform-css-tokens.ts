import { TransformerParams } from './types.js'
import { defaultCSSFormatters } from './css-tokens/index.js'
import { formatList } from './format-list.js'
import { defaultBaseFontSize } from './constants.js'
import { reduceWhitespace } from '@plaited/utils'

const deduplicate = (css: string) => {
  const regex = /((?:.*:host|:host\([^)]*\))[^{\n]*)\{(\s*[\s\S]*?\s*)\}/gm
  const map = new Map<string, Set<string>>()
  let match: RegExpExecArray | null
  while ((match = regex.exec(css)) !== null) {
    const selector = match[1]
    const rule = reduceWhitespace(match[2])
    const set = map.get(selector)
    if (set) {
      set.add(rule)
      continue
    }
    map.set(selector, new Set<string>([rule]))
  }
  return [...map]
    .flatMap(([key, val]) => {
      return [key, '{', ...val, key.startsWith('@') ? '}}' : '}']
    })
    .join('')
}

/**
 * Transforms design tokens into CSS rules with deduplicated selectors on :host.
 * These rules are to be used with design token custom element and applied to
 * it's shadow root's constructable stylesheet.
 * @param {TransformerParams} params - The parameters for the transformation.
 * @returns {string} The transformed CSS rules with deduplicated selectors.
 */
export const transformCssTokens = ({
  tokens,
  baseFontSize = defaultBaseFontSize,
  formatters = defaultCSSFormatters,
  mediaQueries,
  colorSchemes,
  containerQueries,
}: TransformerParams) => {
  const content = formatList({
    tokens,
    allTokens: tokens,
    baseFontSize,
    formatters,
    mediaQueries,
    colorSchemes,
    containerQueries,
  })
  return deduplicate(content)
}
