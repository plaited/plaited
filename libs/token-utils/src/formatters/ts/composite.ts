import { camelCase } from '@plaited/utils'
import { Formatter, AliasValue, FlexToken, GridToken, TypographyToken } from '../../types.js'
import { hasAlias } from '../has-alias.js'
import { resolveTSVar } from '../ts-utils.js'

type CompositeToken = FlexToken | GridToken | TypographyToken

export const composite: Formatter<CompositeToken> = (token, { tokenPath, allTokens }) => {
  const { $value } = token
  if (hasAlias($value)) {
    return `export const ${camelCase(tokenPath.join(' '))} = ${resolveTSVar($value as AliasValue, allTokens)}`
  }
  const toRet = Object.entries($value).map(([key, val]) => `  ${key}: ${resolveTSVar(val, allTokens)},`)
  return [`export const ${camelCase(tokenPath.join(' '))} = {`, ...toRet, '}'].join('\n')
}
