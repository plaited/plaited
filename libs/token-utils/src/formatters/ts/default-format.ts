import { camelCase } from '@plaited/utils'
import { Formatter } from '../../types.js'
import { hasAlias } from '../has-alias.js'
import { isValidAlias, getCssVar } from '../ts-utils.js'

export const defaultFormat: Formatter = (token, { allTokens, tokenPath }) => {
  const { $value } = token
  if (hasAlias($value) && !isValidAlias($value, allTokens)) {
    return ''
  }
  return `export const ${camelCase(tokenPath.join(' '))} = ${getCssVar(tokenPath)}`
}
