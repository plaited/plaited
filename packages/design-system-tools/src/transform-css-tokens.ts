import { formatList } from './format-list.js'
import { DesignTokenGroup, GetFormatters } from './types.js'
const reduceWhitespace = (str: string) => str.replace(/(\s\s+|\n)/g, ' ')

const deduplicate = (css: string) => {
  const regex = /([^\{\n]*)\{(\s*[\s\S]*?\s*)\}/gm
  const map = new Map<string, Set<string>>()
  let match: RegExpExecArray | null
  while ((match = regex.exec(css)) !== null) {
    const selector = match[1]
    const rule = reduceWhitespace(match[2])
    const set = map.get(selector)
    if (set) {
      set.add(rule)
      continue
    }
    map.set(selector, new Set<string>([ rule ]))
  }
  return [ ...map ].flatMap(([ key, val ]) => {
    return [ key, '{', ...val, '}' ]
  }).join('')
}

export const transformCssTokens = async ({
  tokens,
  output,
  baseFontSize,
  formatters,
}: {
  tokens: DesignTokenGroup;
  output: string;
  baseFontSize: number;
  formatters: GetFormatters;
}) => {
  await Deno.mkdir(output, { recursive: true })
  const content = formatList({
    tokens,
    allTokens: tokens,
    baseFontSize,
    formatters,
  })
  deduplicate(content)
  await Deno.writeTextFile(`${output}/tokens.css`, deduplicate(content))
}
