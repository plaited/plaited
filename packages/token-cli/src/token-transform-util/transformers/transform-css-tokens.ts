import fs from 'fs/promises'
import { formatList } from './format-list.js'
import { cssTokens } from '../formatters/index.js'
import { DesignTokenGroup } from '../types.js'
import postcss from'postcss'
import discardDuplicates from 'postcss-discard-duplicates' 

export const transformCssTokens = async ({
  tokens,
  outputDirectory,
  baseFontSize,
}: {
  tokens: DesignTokenGroup,
  outputDirectory: string,
  baseFontSize: number
}) => {
  const output = `${outputDirectory}/css-tokens`
  await fs.mkdir(output, { recursive: true })
  const content = formatList({
    tokens,
    baseFontSize,
    formatters: cssTokens,
  })
  const { css } = await postcss([
    discardDuplicates,
  ]).process(content, { from: undefined, to: '' })
  await fs.writeFile(`${output}/tokens.css`, css)
}
