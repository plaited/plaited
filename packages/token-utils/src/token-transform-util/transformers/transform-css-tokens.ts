import fs from 'fs/promises'
import { formatList } from './format-list.js'
import { cssTokens } from '../formatters/index.js'
import { DesignTokenGroup } from '../../types.js'
import { minify } from 'csso'
import postcss from'postcss'
import combine from 'postcss-combine-duplicated-selectors'
export const transformCssTokens = async ({
  tokens,
  outputDirectory,
  baseFontSize,
}: {
  tokens: DesignTokenGroup,
  outputDirectory: string,
  baseFontSize: number
}) => {
  await fs.mkdir(outputDirectory, { recursive: true })
  const content = formatList({
    tokens,
    allTokens: tokens,
    baseFontSize,
    formatters: cssTokens,
  })
  const { css } = await postcss([
    combine,
  ]).process(content, { from: undefined, to: '' })
  await fs.writeFile(`${outputDirectory}/tokens.css`, minify(css).css)
}
