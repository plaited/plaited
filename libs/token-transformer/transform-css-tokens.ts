import { formatList } from './format-list.ts'
import { DesignTokenGroup, GetFormatters } from '../token-types.ts'
import { combineDuplicatedSelectors, postcss } from '../deps.ts'

export const transformCssTokens = async ({
  tokens,
  outputDirectory,
  baseFontSize,
  formatters,
}: {
  tokens: DesignTokenGroup
  outputDirectory: string
  baseFontSize: number
  formatters: GetFormatters
}) => {
  await Deno.mkdir(outputDirectory, { recursive: true })
  const content = formatList({
    tokens,
    allTokens: tokens,
    baseFontSize,
    formatters,
  })
  const { css } = await postcss([
    combineDuplicatedSelectors,
  ]).process(content, { from: undefined, to: '' })
  await Deno.writeTextFile(`${outputDirectory}/tokens.css`, css)
}
