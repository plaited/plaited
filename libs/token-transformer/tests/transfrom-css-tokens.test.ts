import {
  assertSnapshot,
  beautify,
  describe,
  it,
  resolve,
} from '../../test-deps.ts'
import { transformCssTokens } from '../transform-css-tokens.ts'
import { tokens } from './sample-tokens.ts'
import { defaultCSSFormatters } from '../css-tokens/mod.ts'

const __dirname = new URL('.', import.meta.url).pathname
const outputDirectory = resolve(__dirname, './__tmp__/')

describe('CSS tokens', () => {
  it('transform', async (t) => {
    await transformCssTokens({
      tokens,
      outputDirectory,
      baseFontSize: 20,
      formatters: defaultCSSFormatters,
    })
    const content = await Deno.readTextFile(`${outputDirectory}/tokens.css`)
    assertSnapshot(t, beautify(content, { format: 'css' }))
  })
})
