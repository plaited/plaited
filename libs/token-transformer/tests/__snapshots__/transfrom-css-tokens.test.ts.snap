export const snapshot = {};

snapshot[`CSS tokens > transform 1`] = `
'[data-media="static"]:root {
    --size-x-1: 0.1rem;
    --size-x-2: 0.3rem;
    --size-x-3: 0.6rem;
}

[data-media="mobile"]:root {
    --size-x-1: 0.1rem;
    --size-x-2: 0.4rem;
    --size-x-3: 0.8rem;
}

[data-media="desktop"]:root {
    --size-x-1: 0.1rem;
    --size-x-2: 0.3rem;
    --size-x-3: 0.6rem;
}

[data-media="tv"]:root {
    --size-x-1: 0.2rem;
    --size-x-2: 0.4rem;
    --size-x-3: 0.7rem;
}

:root {
    --size-s-3: 0.6rem;
    --background-color-purple-x-1: #4c00b0;
    --background-color-purple-x-2: #7600bc;
    --background-color-purple-x-3: #8a00c2;
    --border-radius-x-1: 0.15rem;
    --border-radius-x-2: 0.25rem;
    --border-radius-x-3: 0.35rem;
    --font-size-x-1: 0.7rem;
    --font-family-sans-serif: Roboto,sans-serif;
    --font-weight-x-1: 700;
    --letter-spacing-x-1: normal;
    --line-height-x-1: 1;
}'
`;
